packages/debug-test/CDPDebugger.ts

```diff
@@ -0,0 +1,400 @@
+import * as lodash from 'lodash';
+import {testLog} from './TestLog'
+import {
+	shortenIfNeeded,
+	sleep
+} from './TestUtilities';
+import * as WebSocket from 'ws';
+
+type CDPMessageHandler = (data: any) => boolean;
+type EventHandler = (...any) => void;
+
+export type BreakpointInfo = [
+	breakpointId: string,
+	locations: Array<[scriptId: string, lineNumber: number, columnNumber: number]>
+];
+
+export class CDPDebugger {
+
+	constructor(url: string) {
+		this.ws = new WebSocket(url);
+
+		this.wsOpened = new Promise((resolve, reject) => {
+			this.ws.on('open', () => { resolve(); });
+		});
+
+		this.ws.on('message', data => {
+			const obj = JSON.parse(data as string);
+
+			let i = 0;
+			for (; i < this.expectedResponses.length; ++i) {
+				const [description, isExpectedResponse] = this.expectedResponses[i];
+				if (isExpectedResponse(obj)){
+					testLog.message(`received ${description}`);
+					break;
+				};
+			}
+
+			if (i < this.expectedResponses.length) {
+				// found a handler, remove it
+				this.expectedResponses.splice(i, 1);
+			} else {
+				this.logUnexpectedMessage(shortenIfNeeded(data as string, 500));
+			}
+		});
+	}
+
+	public close() {
+		this.ws.close();
+	}
+
+	public on(eventName: string, eventHandler: EventHandler): void {
+		const handlers = this.eventHandlers.get(eventName);
+		if (handlers) {
+			handlers.push(eventHandler);
+		} else {
+			this.eventHandlers.set(eventName, [eventHandler]);
+		}
+	}
+
+	private raise(eventName: string, getArgs: () => any) {
+		const handlers = this.eventHandlers.get(eventName);
+		if (handlers && handlers.length > 0) {
+			testLog.message(`dispatching '${eventName}' event`)
+			// builds args now that we know we have at least one handler
+			const args = getArgs();
+			for (const handler of handlers) {
+				handler(args);
+			}
+		} else {
+			testLog.message(`received '${eventName}' event with no handlers`);
+		}
+	}
+
+	public expectEvent(eventName: string) : Promise<any> {
+		return new Promise((resolve, reject) => {
+			this.on(eventName, eventArgs => resolve(eventArgs));
+		});
+	}
+
+	/**
+	 * Sends a Debugger.enable CDP message (see https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#method-enable).
+	 */
+	public async DebuggerEnable(maxScriptsCacheSize: number = 10000000): Promise<void> {
+		const commandId = this.nextCommandId++;
+		
+		// the actual CDP message
+		const command = `{ "id":${commandId}, "method":"Debugger.enable", "params":{ "maxScriptsCacheSize":${maxScriptsCacheSize} } }`;
```

Comment:
```js
const command = JSON.stringify({
  id: commandId,
  method: 'Debugger.enable',
  params: {
    maxScriptCacheSize: maxScriptsCacheSize
  }
});

Reply:
same in others, prefer creating the object and then stringifying so you get syntax highlighting etc. and it's easier to read than interpolation

Reply:
Replaced with JSON.stringify + object, thanks!